src/gateway/adhoc.cpp

```diff
@@ -42,66 +56,109 @@ adhoc_t::~adhoc_t() {
 }
 
 auto
-adhoc_t::resolve(const partition_t& name) const -> std::vector<asio::ip::tcp::endpoint> {
-    remote_map_t::const_iterator lb, ub;
-
-    auto ptr = m_remotes.synchronize();
-
-    if(!ptr->count(name)) {
-        throw std::system_error(error::service_not_available);
-    }
-
-    std::tie(lb, ub) = ptr->equal_range(name);
-
-    std::uniform_int_distribution<int> distribution(0, std::distance(lb, ub) - 1);
-    std::advance(lb, distribution(m_random_generator));
-
-    COCAINE_LOG_DEBUG(m_log, "providing service using remote actor", attribute_list({
-        {"uuid", lb->second.uuid}
-    }));
-
-    return lb->second.endpoints;
+adhoc_t::resolve(const std::string& name) const -> std::shared_ptr<service_description_t> {
+    return m_remotes.apply([&](const remote_map_t& remotes) {
+        remote_map_t::const_iterator lb, ub;
+        std::tie(lb, ub) = remotes.equal_range(name);
+        if(lb == ub) {
+            throw error_t(error::service_not_available, "service {} not found in gateway", name);
+        }
+
+        // try to find local service
+        auto it = std::find_if(lb, ub, [&](const remote_map_t::value_type& value) -> bool {
+            return value.second->uuid().empty();
+        });
+
+        // local service is found
+        if(it != ub) {
+            COCAINE_LOG_DEBUG(m_log, "providing service using local actor");
+            return it->second;
+        }
+
+        // roll the dice and choose random one
+        std::uniform_int_distribution<int> distribution(0, std::distance(lb, ub) - 1);
+        std::advance(lb, distribution(m_random_generator));
+
+        COCAINE_LOG_DEBUG(m_log, "providing service using remote actor {}", lb->second->uuid());
+
+        return lb->second;
+    });
 }
 
-size_t
-adhoc_t::consume(const std::string& uuid,
-                 const partition_t& name, const std::vector<asio::ip::tcp::endpoint>& endpoints)
-{
-    auto ptr = m_remotes.synchronize();
+auto
+adhoc_t::on_remote_service_exposed(const std::string& uuid,
+                                   const std::string& name,
+                                   unsigned int version,
+                                   const std::vector<asio::ip::tcp::endpoint>& endpoints,
+                                   const io::graph_root_t& protocol) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        remotes.insert({name, std::make_shared<remote_t>(remote_t{uuid, version, endpoints, protocol})});
```

Comment:
It will be there several times. The logic is preserved here and in fact registering same service multiple times is logic violation.